[CORL-645] Add Slack support for v5

nick-funk opened the pull request:
You can find the configuration for Slack channels at `Admin > Configure > Slack`.

You will want to follow the Tettra docs on creating a Slack App to hook this up and test it: 
https://app.tettra.co/teams/coralproject/pages/slack-apps

nick-funk committed:
Create preliminary schema for setting Slack channels

CORL-645

nick-funk committed:
Implement preliminary slack notification using tenant slack channels

CORL-645

nick-funk committed:
Very preliminarily get data loading with FieldArray's for slack channels

CORL-645

nick-funk committed:
Update settings input on schema to allow saving Slack settings to tenant

CORL-645

nick-funk committed:
Filter off UNMODERATED queue events from slack channels

We don't send these to slack through our filters, we only care
about pending, reported, and featured

CORL-645

nick-funk committed:
Include a moderation link in comments pushed to Slack

CORL-645

nick-funk committed:
Hook up proper callback functions for adding/removing slack channels

CORL-645

nick-funk committed:
Add missing translation for Slack navigation item

CORL-645

nick-funk committed:
Update snapshots for preliminary Slack configuration changes

CORL-645

nick-funk committed:
Add some FormField wrappers around slack config elements

Makes the UI appear a little nicer

CORL-645

nick-funk committed:
Merge remote-tracking branch 'origin/master' into feature/CORL-645

nick-funk committed:
Set up slack config to only provide one slack channel

We need to do this until we can get ArrayField's working
in final-form.

CORL-645

nick-funk committed:
Disable the other trigger checkboxes when "All Comments" is checked

CORL-645

nick-funk committed:
Clean up the formatting of Slack messages

CORL-645

nick-funk commented on package.json:
```diff
@@ -274,6 +274,7 @@
     "eventemitter2": "^5.0.1",
     "farce": "^0.2.6",
     "final-form": "4.11.0",
+    "final-form-arrays": "1.1.2",
```
I'm leaving this because we are still using arrays.

I didn't want to hack away all the array logic and make it hard to update to multi-channel support later. This way we can easily update when we get our final-form update in.

wyattjoh replied:
I would suggest removing these from the `package.json` for now, and commenting out parts of the code that included this library until we actually use it.

nick-funk commented on src/core/server/services/slack/publisher.ts:
```diff
@@ -0,0 +1,190 @@
+  const body = commentBody.replace(new RegExp("<br>", "g"), "\n");
+  const moderateLink = createModerationLink(orgURL, comment.id);
+
+  const data = {
```
When building this, do we want to have some kind of templating?
Currently, this will only support English and I could easily see a way to template this with string interpolation and substitution.

wyattjoh replied:
I think in a future iteration we can tie this into i18n, but for now this should meet our requirements.

nick-funk committed:
Add error handling around sending comments to Slack

CORL-645

nick-funk commented on src/locales/en-US/admin.ftl:
```diff
@@ -380,6 +381,27 @@ decisionHistory-rejectedCommentBy = Rejected comment by <Username></Username>
+configure-slack-channel-hookURL-description =
+  Slack provides a channel-specific URL to activate webhook
+  connections. To find the URL for one of your Slack channels,
+  follow these instructions: url
```
These are in flux, as we need to discuss adding Slack details to our docs.

nick-funk committed:
Add links to external Slack setup documentation

CORL-645

wyattjoh commented on src/core/client/admin/routes/Configure/sections/Slack/SlackChannel.tsx:
```diff
@@ -0,0 +1,182 @@
+            parse={parseBool}
+          >
+            {({ input }) => {
+              setAllComments(input.value);
```
Instead of using this, you should use a wrapped field to propagate the field value down:

```
<Field name="allComments">
  {({ input: allCommentsInput }) => (
  ...
  <Field name="reportedComments">
    // use allCommentsInput.value here to disable the input
  </Field>
  ...
  )}
</Field>
```

wyattjoh replied:
See: https://github.com/coralproject/talk/blob/master/src/core/client/admin/routes/Configure/sections/Email/EmailConfigContainer.tsx

nick-funk replied:
Makes sense, fixed!

wyattjoh commented on src/core/server/graph/tenant/schema/schema.graphql:
```diff
@@ -931,6 +931,53 @@ type EmailConfiguration {
+  """
+  channels is the set of Slack channels configured to receive comments
+  """
+  channels: [SlackChannel]
```
This should return instead:

```suggestion
  channels: [SlackChannel!]!
```

wyattjoh commented on src/core/server/graph/tenant/schema/schema.graphql:
```diff
@@ -931,6 +931,53 @@ type EmailConfiguration {
+  """
+  featuredComments is whether this channel will receive featured comments
+  """
+  featuredComments: Boolean
```
```suggestion
  featuredComments: Boolean!
```

Same for other edges in the `SlackChannelTriggers` type.

wyattjoh commented on src/core/server/graph/tenant/schema/schema.graphql:
```diff
@@ -1212,6 +1259,11 @@ type Settings {
+  """
+  slack is the configuration for slack communication
+  """
+  slack: SlackConfiguration @auth(roles: [ADMIN, MODERATOR])
```
```suggestion
  slack: SlackConfiguration! @auth(roles: [ADMIN, MODERATOR])
```

wyattjoh commented on src/core/server/graph/tenant/schema/schema.graphql:
```diff
@@ -3392,6 +3444,49 @@ input CommenterAccountFeaturesInput {
+  """
+  channels are the channels configured for Slack integration
+  """
+  channels: [SlackChannelConfigurationInput]
```
```suggestion
  channels: [SlackChannelConfigurationInput!]
```

wyattjoh commented on src/core/server/services/slack/publisher.ts:
```diff
@@ -0,0 +1,214 @@
+}
+
+function createCommentLink(storyURL: string, commentID: string) {
+  return `${storyURL}?commentID=${commentID}`;
```
As with `src/core/server/services/users/download/download.ts:107`, you should modify this to support story URL's that contain query params:

```suggestion
  const url = new URL(storyURL);
  url.searchParams.set('commentID', commentID);
  return url.href;
```

`URL` is from `import { URL } from 'url';`.

wyattjoh commented on src/core/server/services/slack/publisher.ts:
```diff
@@ -0,0 +1,214 @@
+    comment.revisions.length > 0
+      ? comment.revisions[comment.revisions.length - 1].body
+      : "";
+  const body = commentBody.replace(new RegExp("<br>", "g"), "\n");
```
```suggestion
  const body = commentBody.replace(/<br\/?>/g, "\n");
```

wyattjoh commented on src/core/server/services/slack/publisher.ts:
```diff
@@ -0,0 +1,214 @@
+      redirect: "follow",
+      referrer: "no-referrer",
+      body: JSON.stringify(data),
+    });
```
Why is `cors` specified? This shouldn't apply server side. Also is there a reason `redirect` and `referrer` are set?

wyattjoh commented on src/core/server/services/slack/publisher.ts:
```diff
@@ -0,0 +1,214 @@
+      body: JSON.stringify(data),
+    });
+
+    if (response.status !== 200) {
```
Should check if `response.ok` instead (which is any non 200-299 status code).

nick-funk committed:
Replace form state with wrapped field element

CORL-645

nick-funk committed:
Clean up fetch request sending Slack notifications

CORL-645

nick-funk committed:
Prefer global string replacement over RegEx

CORL-645

nick-funk committed:
Use URL class to construct comment URL's

CORL-645

nick-funk committed:
Require slack configuration in schema

CORL-645

nick-funk committed:
Initialize Slack in fixtures

Also fix up a flaky test that wasn't waiting
on form submission and on-change events.

CORL-645

nick-funk committed:
Merge remote-tracking branch 'origin/master' into feature/CORL-645

nick-funk committed:
Preliminarily fix up styles to match other config pages

CORL-645

nick-funk committed:
Create placeholder add/remove buttons

nick-funk committed:
Convert SlackConfigContainer to FunctionalComponent

CORL-645

nick-funk committed:
Merge remote-tracking branch 'origin/master' into feature/CORL-645

# Conflicts:
#	package.json
#	src/core/client/admin/routes/Configure/Configure.tsx
#	src/core/client/admin/test/configure/__snapshots__/auth.spec.tsx.snap
#	src/core/client/admin/test/configure/general.spec.tsx

nick-funk committed:
Add name field to slack channels

CORL-645

nick-funk committed:
Disable inner fields on Slack channel when not enabled

CORL-645

nick-funk committed:
Improve the delete channel button

CORL-645

nick-funk committed:
Use pureMerge to extract slack channel settings

CORL-645

nick-funk committed:
Do a bit of cleanup on the add channel button

CORL-645
